src/core/tasks_pool.c

```diff
@@ -0,0 +1,122 @@
+#include "tasks_pool.h"
+
+xccl_status_t tasks_pool_init(context *ctx) {
+    ctx->deque_size = (xccl_lib_global_config.tasks_pool_size / LINE_SIZE) + 1;
+    ctx->tasks = (ucc_coll_task_t ***) calloc(2 * ctx->deque_size, sizeof(ucc_coll_task_t **));
+    if (ctx->tasks == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+    ctx->tasks[0] = (ucc_coll_task_t **) calloc(LINE_SIZE, sizeof(ucc_coll_task_t *));
+    if (ctx->tasks[0] == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+    ctx->tasks[ctx->deque_size] = (ucc_coll_task_t **) calloc(LINE_SIZE, sizeof(ucc_coll_task_t *));
+    if (ctx->tasks[ctx->deque_size] == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+    ctx->increase_pool_locks[0] = 0;
+    ctx->increase_pool_locks[1] = 0;
+    ctx->linked_list_locks[0] = 0;
+    ctx->linked_list_locks[1] = 0;
+    ctx->linked_lists[0] = NULL;
+    ctx->linked_lists[1] = NULL;
+    ctx->which_pool = 0;
+    return XCCL_OK;
+}
+
+xccl_status_t increase_tasks(context *ctx, int i) {
+    while (!__sync_bool_compare_and_swap(&(ctx->increase_pool_locks[i / ctx->deque_size]), 0, 1)) {};
```

Comment:
is __sync_bool_compare_and_swap available in gcc only?

Reply:
will be changed to UCS atomics once the next UCX release is out